stats/stats.go

```diff
@@ -55,41 +55,28 @@ func ConditionallyStartStatsServer(addhandlers ...func(mux *http.ServeMux, profi
 }
 
 func StartStatsServerWithPort(startupOpts StartupOptions, addhandlers ...func(mux *http.ServeMux, profiles map[string]string)) {
+	StartCancellableStatsServerWithPort(context.Background(), startupOpts, addhandlers...)
+}
+
+func StartCancellableStatsServerWithPort(ctx context.Context, startupOpts StartupOptions, addhandlers ...func(mux *http.ServeMux, profiles map[string]string)) {
 	// if the env var was provided (i.e., startup is conditional) and the value of that env var is not the expected value, then return and do nothing
 	if startupOpts.EnvVar != "" && os.Getenv(startupOpts.EnvVar) != startupOpts.EnabledValue {
 		return
 	}
 
-	var logLevel zap.AtomicLevel
-	if envLogLevel := os.Getenv("LOG_LEVEL"); envLogLevel != "" {
-		var setLevel zapcore.Level
-		switch envLogLevel {
-		case "debug":
-			setLevel = zapcore.DebugLevel
-		case "warn":
-			setLevel = zapcore.WarnLevel
-		case "error":
-			setLevel = zapcore.ErrorLevel
-		case "panic":
-			setLevel = zapcore.PanicLevel
-		case "fatal":
-			setLevel = zapcore.FatalLevel
-		default:
-			setLevel = zapcore.InfoLevel
-		}
-
-		contextutils.SetLogLevel(setLevel)
-
-	} else if startupOpts.LogLevel != nil {
-		logLevel = *startupOpts.LogLevel
+	if envLogLevel := os.Getenv(contextutils.LogLevelEnvName); envLogLevel != "" {
+		contextutils.SetLogLevelFromString(envLogLevel)
 	}
 
+	// The running instance of the Stats server
+	var server *http.Server
+
 	go RunGoroutineStat()
 
 	go func() {
 		mux := new(http.ServeMux)
 
-		mux.Handle("/logging", logLevel)
+		mux.Handle("/logging", getLoggingHandler(startupOpts))
 
 		addhandlers = append(addhandlers, addPprof, addStats)
```

Comment:
this seems like a race too.. what if the caller depended on the `addhandlers` array behind the slice copy passed in? I think we should copy the array contents before using them in the goroutine

Reply:
great catch here, updated